app/models/conditions_response/membership_status_check.rb

```diff
@@ -0,0 +1,24 @@
+# Revoke membership status if requirements are met
+
+class MembershipStatusCheck < ConditionResponder
+
+  def self.condition_response(condition, log)
+
+    confirm_correct_timing(get_timing(condition), TIMING_EVERY_DAY, log)
+
+    status_updater = MembershipStatusUpdater.instance
+
+    User.members.each do |user|
+
+      if RequirementsForRevokingMembership.requirements_met?(user: user)
+
+        status_updater.revoke_user_membership(user)
+
+        log.record('info', "User #{user.id} (#{user.email}) membership revoked.")
+      end
```

Comment:
This seems awkward since I need to know (for logging purposes) whether or not membership was revoked.  So, I need to call method (`requirements_check`) that will then be called again.   Also, I have to _assume_ that the revoke action does actually success, as I don't have positive confirmation of that.

This could be greatly simplified if we changed the specs for `update_action` and `revoke_update_action` to return `true` or `false`, depending upon whether the action actually occurred.  This could filter up through `check_requirements_and_act` and then be available to higher-level logic such as above.

@weedySeaDragon what do you think?

Reply:
I should have also added:

the implementation of `MembershipStatusUpdater.update_user_status(user)` would mostly just be a call to its `check_requirements_and_act({ user: user })`  but would also return the new status if it's different, and if it's not different, return false.

Then we can refactor `MembershipStatusUpdater.check_user_and_log` to also use that new method.

I can take a run at it if that's helpful.  And if that seems like the right approach.

Reply:
@weedySeaDragon that might work but my first thoughts are:

1. We don't really have a concept of _membership status_.  We have that for ShfApplication, and we have the binary true/false concept for membership.  So, in this case, I'm not sure how we could percolate up an indication of _revoke_ success "final status" with our current structure.

2. Your proposed approach could solve this problem for the case of revoking membership, but it strikes me that this will be a generic problem for many applications of the specification / requirements-check/update-action pattern.  That's why I thought that changing the specs for `update_action` and `revoke_update_action`, to return a boolean, might be a good way to go about this.  

3. Also, for the most part, at the level if this code I really don't care what the actual status of the member is (that is, what is the value of `member` attribute is for this user).   I just want to know whether the action I requested succeeded or not.  On the other hand, I guess there might be situations where I'd want to know some aspect of the _updated_ object.  But even then, that _might_ be the final value of `member` (or `status` if we added that), or it might be some other attribute - so simply reloading the object would suffice to give me that.  So, again. I think being general (as in point 2) is sufficient and accommodates more use cases.

Anyway, please do take a run at it and I'll go with it :)

Reply:
And, yes, (I know what you're thinking ...) we do have an emerging Membership class that wants to be born.

Reply:
So I think the only place I'm not clear is _the boundaries for the responsibilities_.
Here's  how things are now:
  The `MembershipStatusUpdater` is responsible for
  1. updating the membership status for a user
  2. logging any status change to the `<environ>_MembershipStatusUpdater.log`

I'm not clear on what you see as the responsiblity(-ies) of the `MembershipStatusCheck` class.   I am clear about these:
  1. loop through all Users and update their membership status
  2. log any errors encountered 

I sense that you're thinking something more and that's driving your (1) and (2) above.  But I'm not clear on it.  What is the responsibility(-ies) that you see belonging in `MembershipStatusCheck`?  and maybe give me a simple specification if that'll help?

IOW, what responsibility(-ies) requires that `MembershipStatusCheck` knows about the result of the membership update for every single User (that is not already covered by `MembershipStatusUpdater`)?

Reply:
For `MembershipStatusCheck` I would add:

3. Log message for each user whose membership was revoked

I think most (maybe all) of the condition response classes will be logging actions taken.  `Backup` logs general actions (backing up code, backing up DB, posting to AWS, etc.), `DinkursFetch` logs specific actions (event fetched for company N), `MembershipExpireAlert` records each alert email sent to each user, etc.).

For `MembershipStatusUpdater` I would add:

3. Provide confirmation that a requested update (including revocation) happened.  

This latter because a requested update might not succeed, and the requester might then want to take action (such as alerting admins, etc.)

I think that `MembershipStatusUpdater` will not be the only requester that needs this.

Reply:
Here's one of the places that is unclear:

> For MembershipStatusCheck I would add:
>
>  3. Log message for each user whose membership was revoked
>
> I think most (maybe all) of the condition response classes will be logging actions taken. > Backup logs general actions (backing up code, backing up DB, posting to AWS, etc.), DinkursFetch logs specific actions (event fetched for company N), MembershipExpireAlert records each alert email sent to each user, etc.).

`MembershipStatusUpdater` is already logging exactly this.  This is the purpose of that log.  We don't duplicate logging for anything else (for good reason); why do we need to duplicate the logging done by MembershipStatusUpdater?  _insert DRY mention; here there be dragons, etc etc._ ;-)

(Meanwhile today I'll work through the rest of your suggestions with the code.)

Reply:
After pondering this and poking around in the code, I think we need to chat about this in real time.  It seems to me that you're making the `MembershipStatusCheck` class duplicate what `MembershipStatusUpdater` already does.

This is what makes me think that:

> ...This seems awkward since I need to know (for logging purposes) whether or not membership was revoked. So, I need to call method (requirements_check) that will then be called again. Also, I have to assume that the revoke action does actually succeed, as I don't have positive confirmation of that.

I don't see any reason that the `MembershipStatusCheck` needs to interact with the requirements classes at all. It just needs to tell the `MembershipStatusUpdater` to go do whatever needs to be done with each `User` instance.  It is already the responsibility of the `MembershipStatusUpdater` class to deal with any requirements and log everything.

I'm back to trying to get a clear statement of responsibility about the `MembershipStatusCheck` class:  What is the point (responsibility(-ies)) of this class?  

Can we chat about this in real time via Slack or hangouts or after Thursday's status meeting?
I think we'll be able resolve this faster than here.